common/thorhelper/thorread.cpp

```diff
@@ -1272,6 +1274,205 @@ class AlternativeDiskRowReader : public CInterfaceOf<IDiskRowReader>
+
+    if (parquetActivityCtx)
+    {
+        delete parquetActivityCtx;
+    }
+}
+
+IDiskRowStream * ParquetDiskRowReader::queryAllocatedRowStream(IEngineRowAllocator * _outputAllocator)
+{
+    outputAllocator.set(_outputAllocator);
+    return this;
+}
+
+const void * ParquetDiskRowReader::nextRow()
+{
+    return inlineNextRow(
+        [this](size32_t sizeRead, const byte * next)
+        {
+            return next;
+        }
+    );
+}
+
+const void * ParquetDiskRowReader::nextRow(size32_t & resultSize)
+{
+    return inlineNextRow(
+        [this,&resultSize](size32_t sizeRead, const byte * next)
+        {
+            resultSize = sizeRead;
+            return next;
+        }
+    );
+}
+
+const void * ParquetDiskRowReader::nextRow(MemoryBufferBuilder & builder)
+{
+    return inlineNextRow(
+        [this,&builder](size32_t sizeRead, const byte * next)
+        {
+            builder.appendBytes(sizeRead, next);
+            return (const void *)(builder.getSelf() - sizeRead);
+        }
+    );
+}
+
+bool ParquetDiskRowReader::getCursor(MemoryBuffer & cursor)
+{
+    return false;
+}
+
+void ParquetDiskRowReader::setCursor(MemoryBuffer & cursor)
+{
+}
+
+void ParquetDiskRowReader::stop()
+{
+}
+
+void ParquetDiskRowReader::clearInput()
+{
+    eogPending = false;
+}
+
+bool ParquetDiskRowReader::matches(const char * _format, bool _streamRemote, IDiskReadMapping * _mapping)
+{
+    if (!strieq(format, "parquet"))
+        return false;
+    return true; // TO DO add additional check
+}
+
+bool ParquetDiskRowReader::setInputFile(const char * localFilename, const char * logicalFilename, unsigned partNumber, offset_t baseOffset, const IPropertyTree * inputOptions, const FieldFilterArray & expectedFilter)
+{
+    DBGLOG(0, "Opening File: %s", localFilename);
+    parquetFileReader = new parquetembed::ParquetReader("read", localFilename, 50000, nullptr, parquetActivityCtx);
+    auto st = parquetFileReader->processReadFile();
+    if (!st.ok())
+        throw MakeStringException(0, "%s: %s.", st.CodeAsString().c_str(), st.message().c_str());
+    return true;
+}
+
+bool ParquetDiskRowReader::setInputFile(const RemoteFilename & filename, const char * logicalFilename, unsigned partNumber, offset_t baseOffset, const IPropertyTree * inputOptions, const FieldFilterArray & expectedFilter)
+{
+    throwUnexpected();
+}
+
+bool ParquetDiskRowReader::setInputFile(const CLogicalFileSlice & slice, const FieldFilterArray & expectedFilter, unsigned copy)
+{
+    StringBuffer localPath;
+    slice.getURL(localPath, copy);
+
+    // Skip leading .
+    const char * slash = strstr(localPath, "/");
+    // Skip file
+    slash = strstr(slash+1, "/");
+    // Skip ip
+    slash = strstr(slash+1, "/");
+
+    return setInputFile(slash, slice.queryLogicalFilename(), slice.queryPartNumber(), slice.queryStartOffset(), slice.queryFileMeta(), expectedFilter);
+}
+
+template <class PROCESS>
+const void * ParquetDiskRowReader::inlineNextRow(PROCESS processor)
```

Comment:
This probably works, I think it will leak at the moment (because next is never freed).

The documentation is inadequate to say the least.

I suspect for parquet you will need 3 different implementations - very similar to the code for csv.  There are 3 cases:
```
const void *nextRow()
```
This would be the same as this function, but returning next instead of calling processor.  It is used by the engines when they are returning rows to the next stage in the processing

```
const void * nextRow(size32_t & resultSize)
```
These are for the case where the row will be thrown away as soon as it is read (e.g., if filtering, or displaying/counting).  This means the output row can be built up in a temporary buffer which is then reused for subsequent calls.

```
const void * nextRow(MemoryBufferBuilder & builder)
```
This is used for dafilesrv - where the resulting rows are being serialised to any output buffer.  In the csv code this could be implemented by one calling the other.  (I can also see a bug in the last of the csv functions that suggests that it has not been used....)